core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -108,8 +108,59 @@ class TransactionStateManager(brokerId: Int,
     }
   }
 
-  def enablePidExpiration() {
-    // TODO: add producer id expiration logic
+  def enableTransactionalIdExpiration() {
+    scheduler.schedule("transactionalId-expiration", () => {
+      inLock(stateLock) {
+        val now = time.milliseconds()
+
+        val transactionalIdByPartition: Map[Int, mutable.Iterable[String]] = transactionMetadataCache.flatMap { case (_, entry) =>
+          entry.metadataPerTransactionalId.filter { case (_, txnMetadata) =>
+            txnMetadata.txnLastUpdateTimestamp <= now - config.transactionalIdExpirationMs
+          }.map { case (transactionalId, _) =>
+            transactionalId
+          }
+        }.groupBy { transactionalId =>
+          partitionFor(transactionalId)
+        }
+
+        val recordsPerPartition = transactionalIdByPartition
+          .map { case (partition, transactionalIds) =>
+            val deletes: Array[SimpleRecord] = transactionalIds.map { transactionalId =>
+              new SimpleRecord(now, TransactionLog.keyToBytes(transactionalId), null)
+            }.toArray
+            val records = MemoryRecords.withRecords(TransactionLog.EnforcedCompressionType, deletes: _*)
+            val topicPartition = new TopicPartition(Topic.TRANSACTION_STATE_TOPIC_NAME, partition)
+            (topicPartition, records)
+          }
+
+        def removeFromCacheCallback(responses: collection.Map[TopicPartition, PartitionResponse]): Unit ={
+          responses.foreach { case (topicPartition, response) =>
+            response.error match {
+              case Errors.NONE =>
+                inLock(stateLock) {
+                  val idsToRemove = transactionalIdByPartition(topicPartition.partition())
+                  transactionMetadataCache.get(topicPartition.partition)
+                  .foreach{ txnMetadataEntry =>
+                    idsToRemove.foreach { transactionalId =>
+                      txnMetadataEntry.metadataPerTransactionalId.remove(transactionalId)}
+                  }
+                }
+              case _ =>
+                debug(s"writing transactionalId tombstones for partition: ${topicPartition.partition} failed with error: ${response.error.message()}")
+            }
+          }
+        }
+
+        replicaManager.appendRecords(
+          30000, // TODO: what is a reasonable timeout?
+          TransactionLog.EnforcedRequiredAcks,
+          internalTopicsAllowed = true,
+          isFromClient = false,
+          recordsPerPartition,
+          removeFromCacheCallback
+          )
+      }
+    }, period=config.transactionalIdExpirationMs)
```

Comment:
Should we also set the `delay` to `transactionalIdExpirationMs` otherwise the first task will be scheduled immediately and we will have nothing to expire.